### PR TITLE
imp(ibc)!: refactor error handling throughout codebase

### LINKED ISSUE #270: Make core `Error` enums less specific
Follow-up to #255.

Our `Error` enums have too many variants. We should drastically condense them following the ideas described [here](https://github.com/cosmos/ibc-rs/pull/255#issuecomment-1325805412).
```[tasklist]
### Tasks
- [ ] #1320 
- [ ] https://github.com/cosmos/ibc-rs/issues/1249
- [ ] https://github.com/cosmos/ibc-rs/issues/950
```

### LINKED ISSUE #950: Improve error handling when decoding `Any` to `MsgEnvelope`
## Problem Statement

Currently, when converting an incoming IBC message of `Any` type to a `MsgEnvelope`, we are mapping decoding errors to a variant of `RouterError` (e.g. [here](https://github.com/cosmos/ibc-rs/blob/eff17e48da37dfad29e2da9f8993368510badc88/crates/ibc/src/core/msgs.rs#L60-L62)), leading to confusion. Thus, this issue should aim to:

- Fix incorrect mapping
- Utilize more efficient manner to capture the source error

We came across this issue in https://github.com/cosmos/ibc-rs/pull/949 and it is currently blocked on #270. We can address this after refactoring our error handling system.

### LINKED ISSUE #1319: Generalize and consolidate duplicated error variants
## Feature Summary

Duplicated error variants scattered across the ibc-rs codebase should be consolidated as much as possible. These should then live in a more centralized location in the repository, depending on how general they are and on how many ibc-rs crates make use of them. 

## Proposal

One concrete example of this is mentioned in [this](https://github.com/cosmos/ibc-rs/pull/1317#discussion_r1723689924) comment. All error variants that have to do with surfacing decoding errors can be consolidated under one `DecodingError` type that would live in the `ibc-primitives` crate:
```rust
pub enum DecodingError {
    /// invalid identifier: `{0}`
    InvalidIdentifier(IdentifierError),
    /// invalid field: `{0}`
    InvalidFeild(String)
    /// missing field: `{0}`
    MissingFeild(String)
    /// mismatched type URLs: expected `{expected}`, actual `{actual}`
    MismatchedTypeUrls { expected: String, actual: String },
     /// failed to decode with error: `{description}`
    FailedtoDecode { description: String },

    // ... any other relevant variants.
}
```
```[tasklist]
### Tasks
- [ ] https://github.com/cosmos/ibc-rs/pull/1325
```

### LINKED ISSUE #1320: Move host-relevant error variants to a newly-defined `HostError` type
## Feature Summary

Many error variants in ibc-rs are primarily used by hosts and are not relevant to ibc-rs's internals. Such variants should live in one or more dedicated host error types. Such host-relevant error types should live in a workspace that is most-easily accessible by hosts, i.e., in the `ics24-host` workspace.

## Proposal

A new error type specifically for holding host-relevant errors should be defined in `ics24-host`. Any error variants in ibc-rs that exist primarily for hosts' benefit, i.e., they are not used internally by ibc-rs codepaths, or are only used in testing scenarios, should be moved into this host error type.

```rust
pub enum HostError {
    /// some data is invalid
    InvalidData { description: String },
    /// some data is missing
    MissingData { description: String },
    /// failed to update some state
    FailedToUpdateState { description: String },
    /// failed to store some data
    FailedToStoreData { description: String },
    /// failed to parse some data
    FailedToParseData { description: String },
    /// some type that was expected does not exist
    NonexistentType { description: String },
    /// some other classification of error not already covered
    Other { description: String },
}
```

This host error type aims to capture all possible classes of errors that hosts may want to return / propagate, though it will also include an `Other` variant as an option for if a host needs to return an error that is not adequately captured by any of the pre-defined variants.

This `HostError` type is meant to serve as a default error type for hosts to use, and the plan is to have basecoin move away from directly using ibc-rs's error types and use this `HostError` type instead.

Validation, execution, and dispatch trait methods that currently return ibc-rs's top-level `ContextError` type will instead return the new `HostError` type. Additionally, the `ContextError` type will be renamed to `HandlerError` to better capture its semantic meaning. A new `HostError` variant will also be added to it. 

The following are all of the basecoin error variants that need to be captured by the `HostError` type:
<details>
<summary><h3>basecoin error variants</h3></summary>

- [ ] `PacketError` variants
  - [ ] `PacketError::AppModule`
  - [ ] `PacketError::MissingPacketReceipt`? This variant is used in [`ibc-core/ics04-channel/handler`](https://github.com/cosmos/ibc-rs/blob/115969cb5e0ff97df1b493bebb2f2ec0c3aefeb6/ibc-core/ics04-channel/src/handler/recv_packet.rs#L255). Should it be moved to a host-specific error type?
  - [ ] `PacketError::MissingPacketAcknowledgment`

- [ ] `ChannelError` variants
  - [ ] `ChannelError::AppModule`
  - [ ] `ChannelError::MissingCounter`
  - [ ] `ChannelError::FailedToUpdateCounter`
  - [ ] `ChannelError::FailedToStoreChannel`
  - [ ] `ChannelError::NonexistentChannel`

- [ ] `ClientError` variants
  - [ ] `ClientError::InvalidHeaderType`
  - [ ] `ClientError::InvalidRawHeader`
  - [ ] `ClientError::MissingClientState`
  - [ ] `ClientError::MissingUpdateMetaData`
  - [ ] `ClientError::MissingConsensusState`? This variant is used in [`ibc-core/ics02-client/handler`](https://github.com/cosmos/ibc-rs/blob/115969cb5e0ff97df1b493bebb2f2ec0c3aefeb6/ibc-core/ics02-client/src/handler/upgrade_client.rs#L41). Should it be moved to a host-specific error type?

- [ ] `ConnectionError` variants
  - [ ] `ConnectionError::MissingConnection`
  - [ ] `ConnectionError::MissingConnectionCounter`
  - [ ] `ConnectionError::FailedToStoreConnectionIds`
  - [ ] `ConnectionError::FailedToStoreConnectionEnd`
  - [ ] `ConnectionError::FailedToUpdateConnectionCounter`

- [ ] `RouterError` variants
  - [ ] `RouterError::UnknownPort`? This variant is used in multiple places in `ibc-core/ics25-handler/entrypoint`. Should it be moved to a host-specific error type?

- [ ] `TokenTransferError` variants
  - [ ] `TokenTransferError::FailedToParseAccount`

- [ ] `UpgradeClientError` variants
  - [ ] `UpgradeClientError::InvalidUpgradePlan`
  - [ ] `UpgradeClientError::InvalidUpgradeProposal`
  - [ ] `UpgradeClientError::MissingUpgradedClientState`
  - [ ] `UpgradeClientError::MissingUpgradedConsensusState`
  - [ ] `UpgradeClientError::FailedToStoreUpgradePlan`
  - [ ] `UpgradeClientError::FailedToStoreUpgradedClientState`
  - [ ] `UpgradeClientError::FailedToStoreUpgradedConsensusState`
</details>

## Future Considerations

For the time being, the concrete `HostError` defined above will serve as the default type for encapsulating errors from hosts. However, this error type does nothing but encode errors as strings, which might not be ideal or optimal for all hosts' error-handling considerations. We may want to give hosts the option of injecting their own error type so that they have the option of defining and using an error type that is better directly suited to their code base.

This can be done by adding an `Error` associated type to the `ValidationContext` trait. This associated type will need to be constrained via `Into<HandlerError>`. We should also consider what other constraints need to be put on this associated type.

### LINKED ISSUE #1330: Refactor all conversions that convert from raw and `Any` types to return `DecodingError`
## Feature Summary

Currently, many of the `TryFrom` conversions in ibc-rs return a myriad of error types. With the introduction of the `DecodingError` type, it makes sense that any conversions that convert from raw types and `Any` types to domain types should return a `DecodingError`. 

## Proposal

- [x] Migrate all `TryFrom` impls that convert from a raw type or an `Any` type to return a `DecodingError`

### LINKED ISSUE #1339: Merge `PacketError` type into `ChannelError`
## Feature Summary

ibc-rs's top-level error type contains two distinct ICS04 variants, `ChannelError` and `PacketError`. This is a discrepancy from the rest of the variants that exist in this error type, as every other variant captures all the errors that originate from a single module. This discrepancy should be cleaned up by merging `ChannelError` and `PacketError` into a single type.

## Proposal

Merge the `PacketError` type into `ChannelError` by moving the relevant `PacketError` variants into `ChannelError`. We'll opt to do this instead of simply adding an `Packet(PacketError)` variant under `ChannelError` in order to avoid unnecessary enum nesting.

### LINKED ISSUE #1346: Clean up inefficient String variants in error types
## Feature Summary

Application and client-specific errors (like those in 07-tendermint) often get converted to strings and end up in variants like [Other](https://github.com/cosmos/ibc-rs/blob/9d19cc9cfa056ca3d269975d6c5c6271ffc66ccf/ibc-core/ics02-client/types/src/error.rs#L103) or [ClientSpecific](https://github.com/cosmos/ibc-rs/blob/9d19cc9cfa056ca3d269975d6c5c6271ffc66ccf/ibc-core/ics02-client/types/src/error.rs#L97) within lower-level error enums. This approach not only leads to large memory allocations but also makes it difficult for hosts to traverse the error structure to extract or check the specific data they need. Instead, they are forced to parse these strings, which is cumbersome and inefficient.

## Proposal

Clean up and/or consolidate these error variants such that their semantics give a better idea of the underlying error.

### PR DESCRIPTION
**Pull request by seanchen1991**
Closes: #270
Closes: #950
Closes: #1319
Closes: #1320
Closes: #1330
Closes: #1339
Closes: #1346

## Description

The branch of this PR is a feature branch that includes changes from the following PRs:
- [x] #1347
- [x] #1345
- [x] #1343
- [x] #1335
- [x] #1331
- [x] #1325
- [x] #1321
- [x] #1318
- [x] #1317

______

### PR author checklist:
- [X] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [X] Linked to GitHub issue.
- [X] Updated code comments and documentation (e.g., `docs/`).
- [X] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs [hunk 1]
```diff
@@ -21,30 +21,35 @@ pub fn refund_packet_nft_execute(
...
     }
     // mint vouchers back to sender
     else {
         for (i, token_id) in data.token_ids.0.iter().enumerate() {
             let token_uri = data.token_uris.as_ref().and_then(|uris| uris.get(i));
             let token_data = data.token_data.as_ref().and_then(|data| data.get(i));
-            ctx_a.mint_nft_execute(&sender, &data.class_id, token_id, token_uri, token_data)?;
+
+            let _ =
+                ctx_a.mint_nft_execute(&sender, &data.class_id, token_id, token_uri, token_data);
```
Comment (rnbguy) [inline ibc-apps/ics721-nft-transfer/src/handler/mod.rs]: shouldn't this return the error?

File: ibc-clients/ics08-wasm/types/src/consensus_state.rs [hunk 2]
```diff
@@ -54,21 +54,19 @@ impl From<ConsensusState> for Any {
...
         }
         match any.type_url.as_str() {
             WASM_CONSENSUS_STATE_TYPE_URL => decode_consensus_state(&any.value),
-            _ => Err(ClientError::Other {
-                description: "type_url does not match".into(),
-            }),
+            _ => Err(DecodingError::MismatchedResourceName {
+                expected: WASM_CONSENSUS_STATE_TYPE_URL.to_string(),
+                actual: any.type_url.to_string(),
+            })?,
```
Comment (rnbguy) [inline ibc-clients/ics08-wasm/types/src/consensus_state.rs]: I am seeing this pattern often. Maybe we can refactor it as

```rs
if let WASM_CONSENSUS_STATE_TYPE_URL = any.type_url.as_str() {
  Protobuf::<RawConsensusState>::decode(value)
} else {
  Err(...)
}
```

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs [hunk 3]
```diff
@@ -21,30 +21,35 @@ pub fn refund_packet_nft_execute(
...
     ) {
         data.token_ids.as_ref().iter().try_for_each(|token_id| {
-            ctx_a.unescrow_nft_execute(
-                &sender,
-                &packet.port_id_on_a,
-                &packet.chan_id_on_a,
-                &data.class_id,
-                token_id,
-            )
+            ctx_a
```
Comment (rnbguy) [inline ibc-apps/ics721-nft-transfer/src/handler/mod.rs:32]: it seems that the errors are not propagated. `try_for_each` just collects the errors.

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs [hunk 4]
```diff
@@ -21,30 +21,35 @@ pub fn refund_packet_nft_execute(
...
+                .unescrow_nft_execute(
+                    &sender,
+                    &packet.port_id_on_a,
+                    &packet.chan_id_on_a,
+                    &data.class_id,
+                    token_id,
+                )
+                .map_err(NftTransferError::from)
```
Comment (rnbguy) [inline ibc-apps/ics721-nft-transfer/src/handler/mod.rs:40]: probably we can use `.map_err(Into::into)`.

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs (hunk 1 above)
Comment (Farhad-Shabani) [reply ibc-apps/ics721-nft-transfer/src/handler/mod.rs]: Oops! [3ca721d](https://github.com/cosmos/ibc-rs/pull/1350/commits/3ca721d34084e330e8287ba8d44d189c2330dc73)

File: ibc-clients/ics08-wasm/types/src/consensus_state.rs (hunk 2 above)
Comment (Farhad-Shabani) [reply ibc-clients/ics08-wasm/types/src/consensus_state.rs]: Nice suggestion: [91ac57b](https://github.com/cosmos/ibc-rs/pull/1350/commits/91ac57be7636753280ce76df18f1ad2280bbf94d)

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs (hunk 3 above)
Comment (Farhad-Shabani) [reply ibc-apps/ics721-nft-transfer/src/handler/mod.rs:32]: The `try_for_each` actually propagates the error on failures.

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs (hunk 4 above)
Comment (Farhad-Shabani) [reply ibc-apps/ics721-nft-transfer/src/handler/mod.rs:40]: The compiler requires a concrete type for `Into::into`, so using `NftTransferError::from` is more concise IMO.

File: ibc-apps/ics721-nft-transfer/src/handler/mod.rs (hunk 3 above)
Comment (rnbguy) [reply ibc-apps/ics721-nft-transfer/src/handler/mod.rs:32]: Ah. That line break before `else` threw me off.  You are right, it correctly returns the result from `try_for_each`.

**Review by Farhad-Shabani — APPROVED** [review]
Approved! Just one step remains. Apply the changes to `basecoin-rs` and ensure everything looks good there as an example project before the next release.
cc: @rnbguy

**Pull request merged**